File: pset2/caeser/caesar.c
#import <stdio.h>
#import <cs50.h>
#import <string.h>
#import <ctype.h>

string print_encryption(string text, int key);
char encrypt_char(char c, int key);

int main(int argc, string argv[])
{
    if (argc != 2)
    {
        printf("Number of supplied arguments is illegal\n");
        return 1;
    }
    printf("plaintext:");
    string plaintext = get_string();
    printf("ciphertext:%s\n", print_encryption(plaintext, atoi(argv[1])));
    return 0;
}

string print_encryption(string text, int key)
{
    
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        text[i] = encrypt_char(text[i], key);
    }
    return text;
}

char encrypt_char(char c, int key)
{
    if (!isalpha(c))
    {
        return c;
    }
    int base = isupper(c) ? 'A' : 'a';
    int letter = (c % base + key) % 26;
    return base + letter;
}

File: pset6/crypt/caesar.py
from sys import argv
from cs50 import get_string


def main():
    """Main Method"""
    if len(argv) != 2:
        print("Number of supplied arguments is illegal\n")
        exit(1)
    is_number, key = is_number_string(argv[1])
    if not is_number:
        print("Key is not a number\n")
        exit(1)
    plaintext = get_string("plaintext: ")
    ciphertext = encrypt(plaintext, key)
    print("ciphertext: {ciphertext}".format(ciphertext=ciphertext))


def is_number_string(s):
    """
    Checks if provided string `s` is a number
    """
    try:
        n = int(s)
        return True, n
    except ValueError:
        return False, 0


def encrypt(text, key):
    """
    Encrypts a text by the caeser algorithm
    """
    return "".join(map(lambda c: encrypt_char(c, key), text))


def encrypt_char(c, key):
    """
    Encrypts a single char by the caeser algorithm
    """
    # ignoring non alphabetical characters
    if not c.isalpha():
        return c
    # set the base of the alhpabet (upper or lowercase)
    base = ord("A") if c.isupper() else ord("a")
    letter_code = (ord(c) % base + key) % 26
    return chr(base + letter_code)


if __name__ == "__main__":
    main()
